### ISSUE TITLE
 Fetching updates to backend environment: takes a lot of time

### ISSUE THREAD

**Issue by hackmajoris** [aws-amplify/amplify-hosting#3055]
**Which Category is your question related to?**
Amplify Hosting

**Amplify CLI Version**
9.2.1

**What AWS Services are you utilizing?**
* Dynamo Database
* AppSync API
* OpenSearch
* Lambda
* Other

Short story: our project takes almost 1 hour to be deployed.
I'm wondering:
*  why it takes so long to fetch the environment from the cloud: 13+ minutes. (see the pictures)
*  why it takes so long to uploading the files: 10+ minutes. (see the pictures)

Also, the 'assumed role' token which Amplify uses to do the build, expires after 1 hour. We already started to get to that limit, and have no workaround for that.

Any insights are useful. 

Thank you!

![image](https://user-images.githubusercontent.com/3758212/198112625-0d0a7dc1-46cf-47b9-a160-27692d68d384.png)
![image](https://user-images.githubusercontent.com/3758212/198113247-8ebe6b83-3400-4b6e-b24f-6cafad27d208.png)

**Comment by hloriana** [comment]
Hi @hackmajoris can you please provide your App ID so we can investigate why your builds are running for an extended period?

**Comment by hackmajoris** [comment]
@hloriana The app ID is d3gbyqqyggsz5w
I spend the whole day today to investigate this issue. This is what I found.

We use Amplify hooks with JavaScript, and here are some job examples inside hooks:

```
  await forceSetCognitoTriggers(data);
  await createTestUsers(data);
  await updateSsmParams(data);
  await addCustomDataSources(data);
```

Some of them are our logic. Some of them are the post-push fixes which Amplify brakes, like Cognito Triggers, and data source attachments to the AppSync.

Until the version 10 of Amplify, all the hooks files were pushed and pulled to the S3 deployment bucket. You may guess that we need some node packages, like AWS SDK, in order to build those scripts. Locally, node_modules are ignored in the hooks-ignore file and  are not pushed remotely. But remotely, via Amplify Console, the https://github.com/aws-amplify/amplify-hosting/blob/main/scripts/amplifyPush.sh script is used to deploy the app. When it runs, it pushes all the files to the S3 bucket, then does the build. When the build was finished, the hooks were triggered. Because the node_modules at the step where the hooks were triggered were not in place, we had to install them before calling the amplifyPush --simple script. At this step the hooks are executed with no issues, but all those node_modules(20k+ files) ~20mb were pushed and pulled from and to S3 bucket again and again, at each build.

This was consumed a lot of time. The solution I've done so far is this: 

* copied that amplifyPush.sh script in the codebase.
* in the pre-build step, at the amplify.yml, moved the files from amplify/hooks folder in a tmp folder(they are git based)
* modified the amplifyPush.sh script in a way, that after env init, replaced everything from hooks folder(amplify is pulling the files from S3 every time) with that tmp folder and installed the dependencies(the node_modules are ignored in the hooks-ignore file, so they will not be pushed next time)

This could be avoided if this bug https://github.com/aws-amplify/amplify-cli/issues/11021
and this fix https://github.com/aws-amplify/amplify-cli/pull/11179 were already released.

**Comment by hackmajoris** [comment]
Fixed with #11179
